warehouse/packaging/services.py

```diff
@@ -161,3 +167,70 @@ def remove_by_prefix(self, prefix):
             self.s3_client.delete_objects(
                 Bucket=self.bucket_name, Delete={"Objects": keys_to_delete}
             )
+
+
+@implementer(IFileStorage)
+class GCSFileStorage(GenericFileStorage):
+    @classmethod
+    def create_service(cls, context, request):
+        storage_client = request.find_service(name="gcloud.gcs")
+        bucket_name = request.registry.settings["files.bucket"]
+        bucket = storage_client.get_bucket(bucket_name)
+        prefix = request.registry.settings.get("files.prefix")
+
+        return cls(bucket, prefix=prefix)
+
+    def get(self, path):
+        # Note: this is not actually used in production, instead our CDN is
+        # configured to connect directly to our storage bucket. See:
+        # https://github.com/python/pypi-infra/blob/master/terraform/file-hosting/vcl/main.vcl
+        raise NotImplementedError
+
+    def store(self, path, file_path, *, meta=None):
+        path = self._get_path(path)
+        blob = self.bucket.blob(path)
+        blob.upload_from_filename(file_path)
+
+        if meta is not None:
+            blob.metadata = meta
+            blob.patch()
+
+
+@implementer(IFileStorage)
+class DualFileStorage(GenericFileStorage):
+    @classmethod
+    def create_service(cls, context, request):
+        prefix = request.registry.settings.get("files.prefix")
+
+        # Register the primary and secondary service factories
+        request.services.register_factory(
+            ProxyFactory(S3FileStorage.create_service), name="primary-file-backend"
+        )
+        request.services.register_factory(
+            ProxyFactory(GCSFileStorage.create_service), name="secondary-file-backend"
+        )
+
+        # Instanciate the instance of the service and attach the request object
+        service_instance = cls(None, prefix=prefix)
+        service_instance.request = request
+        return service_instance
+
+    def get(self, path):
+        # Note: this is not actually used in production, instead our CDN is
+        # configured to connect directly to our storage bucket. See:
+        # https://github.com/python/pypi-infra/blob/master/terraform/file-hosting/vcl/main.vcl
+        raise NotImplementedError
+
+    def store(self, path, file_path, *, meta=None):
+        # Get the primary and secondary storage services
+        primary_service = self.request.find_service(name="primary-file-backend")
+        secondary_service = self.request.find_service(name="secondary-file-backend")
+
+        # This must succeed
+        primary_service.store(path, file_path, meta=meta)
+
+        # This can fail
+        try:
+            secondary_service.store(path, file_path, meta=meta)
+        except Exception:
+            pass
```

Comment:
Is it worth logging this exception for visibility?

Reply:
Yeah, likely so, I was expecting @ewdurbin to have some thoughts here as well.

Reply:
Probably best to send this to sentry and continue on. That should be doable by issuing at the ERROR level and letting the [logging config](https://github.com/pypa/warehouse/blob/dc120b9c0f8fba9dc98809e68e1b56a1d1bda5e6/warehouse/logging.py#L68) pick it up.

You may also be able to manually create an event by tapping into the registry and finding the sentry client there?